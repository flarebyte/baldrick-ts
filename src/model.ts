export interface PackageJson {
  name: string;
  description: string;
  keywords: string[];
  author: Author | string;
  version: string;
  license: string;
  homepage: string;
  repository: Repository;
  main: string;
  typings: string;
  files: string[];
  engines: Engines;
  scripts: Scripts;
  module: string;
  devDependencies: Dependencies;
  dependencies: Dependencies;
  peerDependencies: Dependencies;
}

export interface PackageKeyStats {
  key: string;
  countItems: number;
  stringLength: number;
}

export interface Author {
  name: string;
  url: string;
}

export interface Dependencies {
  [key: string]: string;
}

export interface Engines {
  node: string;
}

export interface Repository {
  type: string;
  url: string;
}

export interface Scripts {
  [key: string]: string;
}

export enum LicenseType {
  MIT,
  Other,
}

export enum ScaffoldingType {
  TsDx,
  Other,
}

export enum PipelineType {
  Github,
  Other,
}

export enum ProjectType {
  TsLib,
  TsCli,
  Other,
}

export interface CoreProject {
  name: string;
  githubAccount: string;
  licenseType: LicenseType;
  scaffoldingType: ScaffoldingType;
  pipelineType: PipelineType;
  projectType: ProjectType;
}

export interface TodoPriority {
  name: string;
  specialChar: string;
}
export interface Todo {
  name: string;
  description: string;
  priority: TodoPriority;
}

export enum FieldStatus {
  Ok,
  Todo,
  Fixable,
}
export type toFieldStatus = (value: any) => FieldStatus;

export interface PackageJsonStatusConverter {
  name: toFieldStatus;
  description: toFieldStatus;
  keywords: toFieldStatus;
  author: toFieldStatus;
  version: toFieldStatus;
  license: toFieldStatus;
  homepage: toFieldStatus;
  repository: toFieldStatus;
  main: toFieldStatus;
  typings: toFieldStatus;
  files: toFieldStatus;
  engines: toFieldStatus;
  scripts: toFieldStatus;
  module: toFieldStatus;
  devDependencies: toFieldStatus;
  dependencies: toFieldStatus;
  peerDependencies: toFieldStatus;
}

export interface PackageJsonStatus {
  name: FieldStatus;
  description: FieldStatus;
  keywords: FieldStatus;
  author: FieldStatus;
  version: FieldStatus;
  license: FieldStatus;
  homepage: FieldStatus;
  repository: FieldStatus;
  main: FieldStatus;
  typings: FieldStatus;
  files: FieldStatus;
  engines: FieldStatus;
  scripts: FieldStatus;
  module: FieldStatus;
  devDependencies: FieldStatus;
  dependencies: FieldStatus;
  peerDependencies: FieldStatus;
}
