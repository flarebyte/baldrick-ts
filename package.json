{
  "name": "baldrick-ts",
  "description": "Baldrick-ts is a Typescript assistant with a cunning plan that can be used for scaffolding",
  "keywords": [
    "cli",
    "config",
    "continuous-integration"
  ],
  "version": "0.10.0",
  "author": {
    "name": "Olivier Huin",
    "url": "https://github.com/olih"
  },
  "license": "MIT",
  "homepage": "https://github.com/flarebyte/baldrick-ts",
  "repository": {
    "type": "git",
    "url": "https://github.com/flarebyte/baldrick-ts.git"
  },
  "bugs": "https://github.com/flarebyte/baldrick-ts/issues",
  "type": "module",
  "exports": {
    ".": {
      "import": "./dist/src/cli.mjs",
      "default": "./dist/src/cli.mjs",
      "types": "./dist/src"
    },
    "./package.json": {
      "default": "./package.json"
    }
  },
  "files": [
    "dist/src",
    "src"
  ],
  "bin": {
    "baldrick-ts": "dist/src/cli.mjs"
  },
  "engines": {
    "node": ">=14"
  },
  "scripts": {
    "build": "tsc --outDir dist",
    "doc": "npx typedoc --json report/doc.json --pretty src/index.ts && npx baldrick-doc-ts typedoc --json-source report/doc.json",
    "github": "gh repo edit --delete-branch-on-merge --enable-squash-merge",
    "lint:ci": "baldrick lint ci",
    "lint": "baldrick lint check",
    "lint:fix": "baldrick lint fix",
    "md": "baldrick markdown check && baldrick markdown check -s .github/",
    "md:fix": "baldrick markdown fix && baldrick markdown fix -s .github/",
    "prebuild": "yarn reset",
    "ready": "yarn lint && yarn test:cov && yarn md && yarn outdated",
    "reset": "rm -rf dist; rm -rf report",
    "test:ci": "baldrick test ci",
    "test": "baldrick test check",
    "test:cov": "baldrick test cov",
    "test:fix": "baldrick test fix",
    "norm": "npx baldrick-ts generate -f cli -ga 'flarebyte' -ch 'Flarebyte.com' -cy 2021 -l MIT -b baldrick-ts",
    "norm:g": "baldrick-ts generate -f cli -ga 'flarebyte' -ch 'Flarebyte.com' -cy 2021 -l MIT -b baldrick-ts"
  },
  "dependencies": {
    "commander": "^8.3.0",
    "lodash.isequal": "^4.5.0",
    "markdown-table": "^3.0.2",
    "yaml": "^1.10.2"
  },
  "devDependencies": {
    "@types/lodash.isequal": "^4.5.5",
    "baldrick-dev-ts": "0.9.0",
    "typescript": "^4.5.3"
  },
  "peerDependencies": {}
}