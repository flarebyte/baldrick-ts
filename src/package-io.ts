import { PackageJson } from './model';

const simpleCopyPackageJson = (pj: PackageJson): PackageJson => ({
  name: pj.name,
  description: pj.description,
  keywords: pj.keywords,
  author: pj.author,
  version: pj.version,
  license: pj.license,
  homepage: pj.homepage,
  repository: pj.repository,
  main: pj.main,
  typings: pj.typings,
  files: pj.files,
  engines: pj.engines,
  scripts: pj.scripts,
  module: pj.module,
  dependencies: pj.dependencies,
  devDependencies: pj.devDependencies,
  peerDependencies: pj.peerDependencies,
});

export const fromString = (content: string): PackageJson => {
  const results: PackageJson = simpleCopyPackageJson(JSON.parse(content));
  return results;
};

export const toString = (packageJson: PackageJson): string => {
  return JSON.stringify(packageJson, null, 2);
};
