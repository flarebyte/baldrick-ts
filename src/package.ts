import { getNpmScripts } from './markdown-maintenance.js';
import {
  PackageJson,
  CoreProject,
  FieldStatus,
  Todo,
  CustomizedPackageJson,
  minimumNodeVersion,
} from './model.js';
import { copyDependencies, copyScripts } from './package-copy.js';
import { convertPackageToStatus } from './package-status.js';
import { statusToTodo, trimString, trimStringArray } from './utils.js';

const fixme = 'fixme';

const trimPackageJson = (pj: CustomizedPackageJson): CustomizedPackageJson => ({
  description: trimString(pj.description),
  keywords: trimStringArray(pj.keywords),
  author:
    typeof pj.author === 'string'
      ? {
          name: trimString(pj.author),
          url: `https://github.com/${fixme}`,
        }
      : pj.author,
  version: trimString(pj.version),
  scripts: copyScripts(pj.scripts),
  dependencies: copyDependencies(pj.dependencies),
  devDependencies: copyDependencies(pj.devDependencies),
  peerDependencies: copyDependencies(pj.peerDependencies),
});

const normalizeOpenSourcePackage = (
  coreProject: CoreProject,
  customized: CustomizedPackageJson
): PackageJson => {
  const isCli = coreProject.feature.includes('cli');
  const mainExport = isCli ? 'cli' : 'index';
  const bin: { [key: string]: string } = isCli
    ? Object.fromEntries([[coreProject.bin, 'dist/src/cli.mjs']])
    : {};
  return {
    name: coreProject.name,
    description: customized.description,
    keywords: customized.keywords,
    version: customized.version,
    author: customized.author,
    license: coreProject.license,
    homepage: `https://github.com/${coreProject.githubAccount}/${coreProject.name}`,
    repository: {
      type: 'git',
      url: `https://github.com/${coreProject.githubAccount}/${coreProject.name}.git`,
    },
    type: 'module',
    exports: {
      '.': {
        import: `./dist/src/${mainExport}.mjs`,
        default: `./dist/src/${mainExport}.mjs`,
        types: './dist/src',
      },
      './package.json': { default: './package.json' },
    },
    files: ['dist', 'src'],
    bin,
    engines: {
      node: `>=${minimumNodeVersion}`,
    },
    scripts: {
      ...getNpmScripts(coreProject),
    },
    dependencies: customized.dependencies,
    devDependencies: customized.devDependencies,
    peerDependencies: customized.peerDependencies,
  };
};
export const defaultCustomizedPackageJson: CustomizedPackageJson = {
  description: fixme,
  keywords: [],
  version: '0.1.0',
  author: fixme,
  scripts: {},
  dependencies: {},
  devDependencies: {},
  peerDependencies: {},
};

const normalizeOtherPackage = (
  coreProject: CoreProject,
  customized: CustomizedPackageJson
): PackageJson => ({ ...normalizeOpenSourcePackage(coreProject, customized) });

const normalizePackage = (
  coreProject: CoreProject,
  customized: CustomizedPackageJson
): PackageJson =>
  coreProject.license === 'MIT'
    ? normalizeOpenSourcePackage(coreProject, customized)
    : normalizeOtherPackage(coreProject, customized);

export const fixAutomatically = (
  coreProject: CoreProject,
  customized: CustomizedPackageJson
): PackageJson => {
  const trimmed = trimPackageJson(customized);
  const fixed = normalizePackage(coreProject, trimmed);
  return fixed;
};

const keyStatsToTodo = (keyStats: [string, FieldStatus]): Todo => {
  const [key, stats] = keyStats;
  return {
    description: `Key ${key} of package.json`,
    status: statusToTodo(stats),
  };
};

const keepNotOk = (keyStats: [string, FieldStatus]): boolean =>
  keyStats[1] !== 'ok';

export const suggestTasksToDo = (
  coreProject: CoreProject,
  packageJson: PackageJson
): Todo[] => {
  const fixed = fixAutomatically(coreProject, packageJson);
  const packageStatus = convertPackageToStatus(packageJson, fixed);
  const results = Object.entries(packageStatus)
    .filter(keepNotOk)
    .map(keyStatsToTodo);
  return results;
};
