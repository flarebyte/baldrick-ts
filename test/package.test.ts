import { CustomizedPackageJson } from '../src/model.js';
import { fixAutomatically, suggestTasksToDo } from '../src/package.js';
import { fromString } from '../src/package-io.js';
import { libCoreProject } from './fixture-core-project.js';

describe('Suggestions', () => {
  const customizedPackageJson: CustomizedPackageJson = {
    description: 'some description of the project',
    keywords: ['testing'],
    version: '0.7.0',
    author: {
      name: 'Olivier',
      url: 'http://mywebsite.com',
    },
    scripts: {},
    devDependencies: { typescript: '^4.5.3' },
    dependencies: { commander: '^8.3.0' },
    peerDependencies: {},
  };
  const refPackageJson = fixAutomatically(
    { ...libCoreProject },
    customizedPackageJson
  );
  const ref = fromString(JSON.stringify(refPackageJson));

  it('Fix automatically', () => {
    const actual = fixAutomatically(libCoreProject, ref);
    expect(actual).toMatchSnapshot();
  });

  it('Suggest tasks to do', () => {
    const actual = suggestTasksToDo(libCoreProject, ref);
    expect(actual).toMatchSnapshot();
  });
});
