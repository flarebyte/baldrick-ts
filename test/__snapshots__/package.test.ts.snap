// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Suggestions Fix automatically 1`] = `
Object {
  "author": Object {
    "name": "Olivier",
    "url": "http://mywebsite.com",
  },
  "bin": Object {},
  "bugs": "https://github.com/mycompany/scratchbook/issues",
  "dependencies": Object {
    "commander": "^8.3.0",
  },
  "description": "some description of the project",
  "devDependencies": Object {
    "typescript": "^4.5.3",
  },
  "engines": Object {
    "node": ">=14",
  },
  "exports": Object {
    ".": Object {
      "default": "./dist/src/index.mjs",
      "import": "./dist/src/index.mjs",
      "types": "./dist/src",
    },
    "./package.json": Object {
      "default": "./package.json",
    },
  },
  "files": Array [
    "dist/src",
    "src",
  ],
  "homepage": "https://github.com/mycompany/scratchbook",
  "keywords": Array [
    "testing",
  ],
  "license": "MIT",
  "name": "scratchbook",
  "peerDependencies": Object {},
  "repository": Object {
    "type": "git",
    "url": "https://github.com/mycompany/scratchbook.git",
  },
  "scripts": Object {
    "build": "yarn reset;tsc --outDir dist",
    "lint": "baldrick lint check",
    "lint:ci": "baldrick lint ci",
    "lint:fix": "baldrick lint fix",
    "norm": "npx baldrick-ts generate -f lib -ga 'mycompany' -ch 'Great Company' -cy 2020 -l MIT -b scratch",
    "norm:g": "baldrick-ts generate -f lib -ga 'mycompany' -ch 'Great Company' -cy 2020 -l MIT -b scratch",
    "ready": "yarn lint && yarn test:cov && yarn outdated",
    "reset": "rm -rf dist; rm -rf report",
    "test": "baldrick test check",
    "test:ci": "baldrick test ci",
    "test:cov": "baldrick test cov",
    "test:fix": "baldrick test fix",
  },
  "type": "module",
  "version": "0.7.0",
}
`;

exports[`Suggestions Suggest tasks to do 1`] = `Array []`;
